presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoStatement.java

```diff
@@ -247,9 +247,10 @@ final boolean internalExecute(String sql)
             WarningsManager warningsManager = new WarningsManager();
             currentWarningsManager.set(Optional.of(warningsManager));
             resultSet = new PrestoResultSet(client, maxRows.get(), progressConsumer, warningsManager);
+            String updateType = client.isFinished() ? client.finalStatusInfo().getUpdateType() : client.currentStatusInfo().getUpdateType();
```

Comment:
This is not exactly a synchronization problem as multiple threads are not involved in this case. Statement object is used only by single thread and is not thread-safe in general. It is more of the getter method client.currentStatusInfo() throwing exception if state is not running anymore. Since all we need is updateType, used the finalStatusInfo() to get the QueryStatusInfo avoiding the exception instead of adding another getter method that does not throw exception. We ran into this error because we ran 100+ java processes running same jdbc code and there was so much context switch that between client = connection().startQuery(..) and getUpdateType() the client had finished.

Reply:
There is the thread executing this method and the worker thread in `PrestoResultSet` that is using the client. Even with this check, there is a small race between the `isFinished()` call the call to fetch current status (it could finish in that time).

I think the easiest solution is to simply remove the check in `currentStatusInfo()` so that it is always valid to call that method. That would allow us to simplify a few calls that check `isRunning()` to determine which method to call. The check for `finalStatusInfo()` should stay, as it can be important for the caller to know that this is the final status (and thus will never change).

Note that `isRunning()` is actually the proper check with the code today, as `isFinished()` does not include error states. (this is one of the more confusing parts about the client API)